Fix incorrect version-specific channel routing in NodeGraph visualiser

mjftw opened the pull request:
## Motivation

### Context

I was using EventCatalog on my own catalogue and wanted to add an example to show how a service might dual publish the same event on two different major versions, to two different channels (different Kafka topics in this case).

I found that the catalogue was not rendering how I had expected it to. I thought it was an EventCatalog bug, so I decided to see if I could reproduce it in the examples and fix it rather than putting up a Github Issue and hoping someone else would fix it.

### Issue as seen in my Event Catalogue

Here's what it looked like in my service's .mdx frontmatter:

```yaml
---
id: example_orders_service
name: Example Orders Service
version: 1.0.0
sends:
  - id: event:example.order.created
    version: 1.x
    to:
      - id: channel:example.order.created
        version: 1.x
  - id: event:example.order.created
    version: 2.x
    to:
      - id: channel:example.order.created
        version: 2.x
---
```

And in the UI:

<img width="1906" height="694" alt="Screenshot 2025-12-24 at 16 13 30 (1)" src="https://github.com/user-attachments/assets/c13899ea-64e6-4c27-9aee-3a71a7b1b062" />
<img width="2474" height="1030" alt="Screenshot 2025-12-24 at 16 21 44 (1)" src="https://github.com/user-attachments/assets/8c66bb6e-81fd-43e4-b637-7c81dc952ec9" />

In my Service frontmatter version 1.x of the event should go to 1.x of the channel, and 2.x event to 2.x channel, but that's not what the visualiser shows, so I think this is a bug.

This PR fixes incorrect event-to-channel routing in NodeGraph visualisation when a service sends multiple versions of the same event to different version-specific channels.

Here is a screenshot taken from my catalogue after using the fixed version of `@eventcatalog/core` (changes in this PR), showing the issue resolved.

<img width="1029" height="265" alt="Screenshot 2025-12-30 at 09 45 38" src="https://github.com/user-attachments/assets/ff938330-1fd7-40f4-9163-66ec917903bc" />

More info on the bug and fix below.

### The Problem - reproduced in EventCatalog source tree

I was able to reproduce the issue in the EventCatalog examples by setting up a similar scenario.

When a service is configured to send different versions of an event to different channels:

```yaml
sends:
  - id: InventoryAdjusted
    version: 0.0.1
    to:
      - id: inventory-adjusted-channel
        version: 0.0.1
  - id: InventoryAdjusted
    version: 1.0.1
    to:
      - id: inventory-adjusted-channel
        version: 1.0.1
```

The NodeGraph visualisation would incorrectly show **both** event versions routing to the **same** channel (whichever was found first), rather than each version routing to its correct version-specific channel.

### Screenshots

**Before (Bug):** InventoryAdjusted v1.0.1 incorrectly routes to `inventory-adjusted-channel` v0.0.1

<img width="1962" height="1120" alt="Screenshot 2025-12-24 at 17 29 18" src="https://github.com/user-attachments/assets/72a0cb0c-96d3-487f-85bf-7e2f58551515" />

**After (Fix):** InventoryAdjusted v1.0.1 correctly routes to `inventory-adjusted-channel` v1.0.1

<img width="2592" height="1156" alt="Screenshot 2025-12-24 at 17 57 08" src="https://github.com/user-attachments/assets/9c1382b3-0fa4-4111-b4d8-17ec8e370d08" />

### Root Cause

The `.find()` method in `message-node-graph.ts` and `services-node-graph.ts` was only matching by `id`, ignoring the `version` field:

```typescript
// Before: Only matched on id, ignoring version
const config = sends.find((send) => send.id === message.data.id);
```

This caused the first matching entry to be returned regardless of version.

### The Fix

1. Added a generic `versionMatches()` utility function to `utils/collections/util.ts` that correctly handles version matching with comprehensive pattern support:
   - **Exact versions** (1.0.0 matches 1.0.0)
   - **Semver ranges** (^1.0.0, ~1.2.0, >=1.0.0 <2.0.0) using the `semver` package
   - **X-patterns** (1.x, 1.2.x) with correct boundary checks
   - **`'latest'` keyword** (matches any version)
   - **`undefined` version** (matches any version)

2. Updated the node graph specific `versionMatches()` helper in `utils/node-graphs/utils/utils.ts` to use the generic utility whilst handling graph-specific edge cases (`undefined` and `'latest'`)

3. Updated `getProducersOfMessage()` and `getConsumersOfMessage()` in `utils/collections/services.ts` to use the new `versionMatches()` utility, replacing direct `semver.satisfies()` calls

4. Updated all relevant `.find()` calls in `message-node-graph.ts` (6 locations) and `services-node-graph.ts` (2 locations) to include version matching:

```typescript
// After: Matches on both id AND version, with semver/x-pattern support
const config = sends.find(
  (send) => send.id === message.data.id && versionMatches(send.version, message.data.version)
);
```

## Semver and X-Pattern Support

The implementation now supports semver ranges and x-pattern matching!

Services can now use flexible version patterns in their `sends` and `receives` configurations:

```yaml
# Caret ranges - accepts compatible versions
sends:
  - id: OrderCreated
    version: ^1.0.0  # Matches 1.0.0, 1.2.3, 1.9.9, but not 2.0.0

# Tilde ranges - accepts patch-level changes
receives:
  - id: PaymentProcessed
    version: ~1.2.0  # Matches 1.2.0, 1.2.5, but not 1.3.0

# X-patterns - wildcard matching
receives:
  - id: InventoryAdjusted
    version: 1.x     # Matches any 1.x.x version
  - id: OrderCreated
    version: 1.2.x   # Matches 1.2.0, 1.2.5, but not 1.3.0
```

This enables more flexible version management strategies while maintaining correct graph visualisation and routing.

When a pattern matches multiple catalogue versions, the latest matching version is resolved — keeping the graph clean and unambiguous.

## Example Catalogue Demonstrates Version Patterns

The example catalogue now includes a minimal demonstration of the three version matching patterns:

Run the catalogue:
```bash
pnpm run start:catalog
```

Navigate to the service visualisers to see version pattern matching in action:
- **http://localhost:3000/visualiser/services/InventoryService/0.0.2** - Sends `InventoryAdjusted` v1.0.0 (exact match)
<img width="990" height="450" alt="Screenshot 2026-02-20 at 18 52 27" src="https://github.com/user-attachments/assets/500b6628-273a-4f46-b676-1fb250847000" />


- **http://localhost:3000/visualiser/services/NotificationService/0.0.2** - Receives `InventoryAdjusted` ^1.0.0 (resolves to latest match)
<img width="1072" height="607" alt="Screenshot 2026-02-20 at 18 53 08" src="https://github.com/user-attachments/assets/e9e75e01-e5cf-48bf-874a-f6d4d44d33b1" />

- **http://localhost:3000/visualiser/services/OrdersService/0.0.3** - Receives `InventoryAdjusted` 1.x (resolves to latest match)
<img width="1145" height="334" alt="Screenshot 2026-02-20 at 18 53 32" src="https://github.com/user-attachments/assets/8006b99b-207a-4567-98bd-7203942e7ece" />


## Tests Added

### Unit Tests
- **`eventcatalog/src/utils/__tests__/node-graphs/utils/utils.spec.ts`** - Tests for graph-specific `versionMatches()` wrapper including delegation to generic utility
- **`eventcatalog/src/utils/__tests__/collections/util.spec.ts`** - Comprehensive tests for generic `versionMatches()` utility covering exact versions, semver ranges (^, ~, complex), x-patterns, latest keyword, and boundary edge cases

### Integration Tests
- **`eventcatalog/src/utils/__tests__/messages/node-graph.spec.ts`** - Integration tests verifying:
  - Version-specific channel routing (original bug fix)
  - Producer/consumer matching with caret ranges (^1.0.0)
  - Tilde range matching (~1.2.0)
  - X-pattern matching at major (1.x) and minor (1.2.x) levels
  - Semver patterns with channel configurations
  - Explicit multi-version receives (service listing multiple specific versions)

boyney123 commented:
Thanks for finding this @mjftw and writing the details response helped me understand. Looks like a bug for sure.

Your questions

> Currently the changes will only allow matching on the exact version if specified (e.g. 1.0.0 matches 1.0.0) - would it be better to also allow semver and x matching? (e.g. ^1.0.0 or 1.x)

Yeah I think we should, we match on semver and x matching for other references when we use versions I believe, what do you think? could be a good addition, are you OK to add that?

Also think some minor conflicts in your PR, sorry it took me so long to get back on this.

mjftw commented:
@boyney123 I've now updated this PR to drop the functionality added around getting multiple matches as suggested (this is much simpler too), and I've rebased on main.
I've confirmed everything is working, but I'd suggest you checkout the branch and run the manual testing instructions yourself just to be sure 🙂 

This PR is now hopefully good to go 🤞🏽

mjftw commented:
> Great thanks for the effort here @mjftw !
> 
> Just a quick one try run `npm run format` which should format the code for the linter.

No problem, done!

mjftw commented:
Ack reading through the PR I think some things got messed up during the rebase following the move to monorepo setup.
Will fix.
